Add AXI PWM generator driver to mainline - MMIO regmap version

threexc opened the pull request:
This is a continuation of the work in https://github.com/analogdevicesinc/linux/pull/2363, with the following changes:

- use of MMIO regmap and sanity checking of the clk rate as discussed in #2363
- use of dev_err_probe in the probe and setup functions
- Some cleanup (mostly with checkpatch and review from @dlech)

I've copied the previous reviewer list and added @pdp7. As before, the axi-pwmgen driver is related to the SPI offload upstreaming that @dlech has been doing in https://github.com/analogdevicesinc/linux/pull/2341 as both are required to upstream drivers like the ad400x.

Checkpatch still throws a warning for a handful of the error messages due to line length, but I have left them intact so they're searchable.

nunojsa commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,236 @@
+	u32 val;
+
+	ret = regmap_write(regmap, AXI_PWMGEN_REG_SCRATCHPAD, AXI_PWMGEN_TEST_DATA);
+	if (ret < 0)
```
values lower than 0 have no meaning in regmap... Just do `if (ret)`

threexc replied:
Done.

nunojsa commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,236 @@
+				     "failed to read PWM count register\n");
+
+	if (pwm->chip.npwm > AXI_PWMGEN_NPWM)
+		return -EINVAL;
```
I would argue about `-EINVAL`. This means either there's something wrong with the core (so likely EIO) or the core changed and the driver was left behind...

threexc replied:
Changed to -EIO.

nunojsa commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,236 @@
+	for (idx = 0; idx < pwm->chip.npwm; idx++) {
+		ret = regmap_write(regmap, AXI_PWMGEN_CHX_PERIOD(idx), 0);
+		if (ret < 0)
+			return dev_err_probe(dev, -EIO,
```
don't ignore `ret` and use -EIO... Just use `ret`. Ditto for all the other places. In fact, I would not print any error message for failed remap stuff. If regmap fails, then something seriously wrong is going on with the device

threexc replied:
Mostly done - I left an error message for devm_regmap_init_mmio and devm_clk_get_enabled. Should those change too?

nunojsa replied:
> Should those change too?

I typically don't add as my naive assumption is that the error should be printed in the APIs but up to you...

threexc replied:
Alright, I'll leave them in for now.

nunojsa commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,236 @@
+	if (ret < 0)
+		return dev_err_probe(dev, -EIO, "failed to update configuration register\n");
+
+	return 0;
```
I would just do `return regmap_update_bits()`.

threexc replied:
Done.

nunojsa commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,236 @@
+	struct axi_pwmgen *pwm = to_axi_pwmgen(chip);
+	unsigned long clk_rate_hz = clk_get_rate(pwm->clk);
+	struct regmap *regmap = pwm->regmap;
+	size_t ch = device->hwpwm;
```
why `size_t`?

threexc replied:
Thanks for the catch. Changed to unsigned int to match the pwm_device struct.

dlech commented on Documentation/devicetree/bindings/pwm/adi,axi-pwmgen.yaml:
```diff
@@ -0,0 +1,48 @@
+
+properties:
+  compatible:
+    const: adi,axi-pwmgen-v1
```
Versions in compatible strings for Xylinx FPGA IP blocks tend to looks like `-1.00.a` rather than `-v1` (e.g. `adi,axi-dmac-1.00.a`, `adi,axi-clkgen-2.00.a` and others already upstream)

threexc replied:
changed it to axi-pwmgen-1.00.a

dlech commented on drivers/pwm/Kconfig:
```diff
@@ -98,6 +98,18 @@ config PWM_ATMEL_TCB
+	help
+	  This enables support for the Analog Devices AXI PWM generator.
+
+	  This is a single output configurable PWM generator with variable
```
It can have up to 4 channels, so we should probably not mention "single output" here.

threexc replied:
Removed "single output" from the description.

dlech commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,226 @@
+ * Analog Devices AXI PWM generator
+ *
+ * Copyright 2024 Analog Devices Inc.
+ * Copyright 2024 Baylibre SAS Inc.
```
```suggestion
 * Copyright 2024 Baylibre SAS
```

threexc replied:
Fixed! Thanks for catching

dlech commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,226 @@
+		return ret;
+
+	if (val != AXI_PWMGEN_TEST_DATA)
+		return ret;
```
`ret` will always be 0 here, so it would return success even though this should be considered a failure. I think keeping the return from the original driver here is appropriate.

```suggestion
		return dev_err_probe(dev, -EIO,
				     "failed to access the device registers\n");
```

threexc replied:
Oof, thanks for catching that... I knew there was something wrong with that line but got distracted :)

dlech commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,226 @@
+	if (ret)
+		return ret;
+
+	if (pwm->chip.npwm > AXI_PWMGEN_NPWM)
+		return -EIO;
```
This probably doesn't need to be an error, but rather a warning. We can still use 4 channels even if the IP core reports it supports more than 4 channels.
```suggestion
	if (pwm->chip.npwm > AXI_PWMGEN_NPWM) {
		dev_warn(dev, "driver is limited to %d channels\n but hardware reported %u\n",
			AXI_PWMGEN_NPWM, pwm->chip.npwm);
		pwm->chip.npwm = AXI_PWMGEN_NPWM
	}
```

threexc replied:
Done

dlech commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,226 @@
+}
+
+static const struct of_device_id axi_pwmgen_ids[] = {
+	{ .compatible = "adi,axi-pwmgen" },
```
Needs to match device tree bindings.

threexc replied:
Changed to axi-pwmgen-1.00.a to match the other comment

threexc commented:
Pushed another minor change to fix the binding example.

threexc commented:
With both reviews I think this is good to go now, but I'll leave it open until next week in case there's any more feedback.

nunojsa commented on drivers/pwm/Kconfig:
```diff
@@ -98,6 +98,18 @@ config PWM_ATMEL_TCB
 
+config PWM_AXI_PWMGEN
+	tristate "Analog Devices AXI PWM generator"
+	depends on HAS_IOMEM
```
Ups... With the regmap move we want `select REGMAP_MMIO`

threexc replied:
Done

nunojsa commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,229 @@
+	bool			ch_enabled[AXI_PWMGEN_NPWM];
+};
+
+struct regmap_config axi_pwm_regmap_config = {
```
`static const`

threexc replied:
Done

nunojsa reviewed (commented):
Some more comments that I just realized now... With that fixed, I think we are good to go

nunojsa commented on drivers/pwm/Kconfig:
```diff
@@ -98,6 +98,18 @@ config PWM_ATMEL_TCB
 
+config PWM_AXI_PWMGEN
+	tristate "Analog Devices AXI PWM generator"
+	depends on REGMAP_MMIO
```
just select it... try `git grep REGMAP_MMIO` and see what you get :)

threexc replied:
Done

nunojsa reviewed (approved):
With my comment addressed...

threexc commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,229 @@
+#define AXI_PWMGEN_CHX_PERIOD(ch)	(AXI_PWMGEN_CH_PERIOD_BASE + (12 * (ch)))
+#define AXI_PWMGEN_CHX_DUTY(ch)		(AXI_PWMGEN_CH_DUTY_BASE + (12 * (ch)))
+#define AXI_PWMGEN_CHX_OFFSET(ch)	(AXI_PWMGEN_CH_OFFSET_BASE + (12 * (ch)))
+#define AXI_PWMGEN_TEST_DATA		0x5A0F0081
```
@nunojsa does this AXI_PWMGEN_TEST_DATA value have any significance? Upstream is wondering and I got here after it was added :)

nunojsa replied:
Hmm no idea... My feeling is that it's just a random value to write in the scratch pad register and read it back. Don't think it has any special meaning (at least I can't find anything on the wiki page about it).

threexc commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,218 @@
+		return -EINVAL;
+	}
+
+	if (state->enabled) {
+		ret = regmap_read(regmap, AXI_PWMGEN_CHX_PERIOD(ch), &cnt);
+		if (ret)
+			return ret;
+	} else {
+		AXI_PWMGEN_CHX_PERIOD(ch) == 0;
+	}
```
Oof, thanks for catching that. Not sure how I left the AXI_PWMGEN_CHX_PERIOD(ch) == 0; line...

threexc commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,218 @@
+
+struct axi_pwmgen_ddata {
+	struct pwm_chip	chip;
+	struct clk *clk;
```
clk is used to derive clk_rate_hz. Is the clk structure hidden somewhere in one of these other structs that I'm not seeing? Otherwise, I don't know how to get the value.

threexc commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,218 @@
+		return ret;
+
+	/* Enable the core */
+	return regmap_update_bits(regmap, AXI_PWMGEN_REG_CONFIG, AXI_PWMGEN_RESET, 0);
```
Maybe I'm misunderstanding what's happening, but it looks like this is writing a 0 into the reset, which is already 0 by default?

0x0004 	0x0010 	REG_RSTN 	Reset and load values
	                        [1] 	LOAD_CONFIG 	WO 	0x0 	Loads the new values written in the config registers.
                                [0] 	RESET 	                RW 	0x0 	Reset, default is (0x0).

threexc replied:
The formatting doesn't carry over well...

dlech replied:
It isn't exactly clear to me from the docs what the intention of how this is supposed to work. But I think that writing 1 would put the IP core into reset and writing 0 takes it out of reset. So writing 0 when already out of reset is probably fine (has no side-effects). @acostina could probably clarify if I am correct or not.

In any case, it would be good to test that it doesn't have any unintended side effects (e.g unbind the driver while PWM is running and bind it again and see if it glitches).

nunojsa replied:
> It isn't exactly clear to me from the docs what the intention of how this is supposed to work. But I think that writing 1 would put the IP core into reset and writing 0 takes it out of reset.

Likely that... Well, we can also attach a scope (if there's some test point where to attach it) and be sure about it :)

threexc commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,218 @@
+		return ret;
+
+	duty_cnt = mul_u64_u64_div_u64(state->duty_cycle, ddata->clk_rate_hz, NSEC_PER_SEC);
+	ret = regmap_write(regmap, AXI_PWMGEN_CHX_DUTY(ch), duty_cnt);
```
Good point. I'm not sure what's best in the case of a 0% duty cycle, though - it looks like some other pwm drivers (e.g. drivers/pwm/pwm-samsung.c) don't support 0% and increment the variable to avoid it, while others set it to 0 under certain conditions.

threexc replied:
Currently, my update looks like:

`          duty_cnt = mul_u64_u64_div_u64(state->duty_cycle, ddata->clk_rate_hz, NSEC_PER_SEC);
~         if (duty_cnt > UINT_MAX)
+                 duty_cnt = UINT_MAX;
+ 
+         ret = regmap_write(regmap, AXI_PWMGEN_CHX_DUTY(ch), state->enabled ? duty_cnt : 0);
          if (ret)
                  return ret;
`

dlech replied:
Allowing 0% duty cycle makes sense to me.

threexc commented on drivers/pwm/Kconfig:
```diff
@@ -98,6 +98,19 @@ config PWM_ATMEL_TCB
+	tristate "Analog Devices AXI PWM generator"
+	depends on HAS_IOMEM
+	select REGMAP_MMIO
+	help
```
That's a cool idea. I'll keep that in mind as a possible future improvement.

nunojsa replied:
I think like this is pretty much what exists today. But it yes, would be a cool idea. Not sure how easy to have it in a sane way

threexc commented:
I've pushed an update to the branch covering almost everything, except for the question for @acostina and the question about Kconfig dependency (which I want to look more into).

nunojsa commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,224 @@
+
+static void devm_clk_rate_exclusive_put(void *data)
+{
+	struct clk *clk = data;
```
pass data directly to he call...

threexc replied:
Done

nunojsa commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,224 @@
+	ddata->clk_rate_hz = clk_get_rate(clk);
+	if (!ddata->clk_rate_hz || ddata->clk_rate_hz > NSEC_PER_SEC)
+		return dev_err_probe(&pdev->dev, -EINVAL,
+				     "failed to get clock rate\n");
```
I would rephrase it for "Invalid clock rate" and maybe print it out?

threexc replied:
You're right, that's a better phrasing. Changed it.

nunojsa commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,224 @@
+module_platform_driver(axi_pwmgen_driver);
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Sergiu Cuciurean <sergiu.cuciurean@analog.com>");
```
nit: well, at this point you could very well add yourself as an author since you added significant changes. Obviously up to you :)

threexc replied:
Good idea :)

nunojsa reviewed (commented):
No much to add :)

dlech commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,223 @@
+	ddata->clk_rate_hz = clk_get_rate(clk);
+	if (!ddata->clk_rate_hz || ddata->clk_rate_hz > NSEC_PER_SEC)
+		return dev_err_probe(&pdev->dev, -EINVAL,
+				     "Invalid clock rate\n");
```
```suggestion
				     "Invalid clock rate: %lu\n", ddata->clk_rate_hz);
```

I like Nuno's suggestion of printing the rate here.

dlech commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,223 @@
+	return regmap_update_bits(regmap, AXI_PWMGEN_REG_CONFIG, AXI_PWMGEN_RESET, 0);
+}
+
+static void devm_clk_rate_exclusive_put(void *data)
```
This function should use the same prefix as other functions in this file.
```suggestion
static void axi_pwmgen_clk_rate_exclusive_put(void *data)
```

threexc commented:
Added a change to print out the clk rate if it's invalid and changed the prefix for the _put function. Also pushed another change just now to remove the max register after David pointed out this PR: https://github.com/analogdevicesinc/hdl/pull/1213

The hard limit on number of PWMs was already removed previously.

threexc commented:
Pushed another version that checks the IP core version by reading AXI_PWMGEN_REG_CORE_VERSION. Based on https://github.com/analogdevicesinc/linux/pull/2404

dlech commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,233 @@
+	if (val != AXI_PWMGEN_TEST_DATA)
+		return dev_err_probe(dev, -EIO, "failed to access the device registers\n");
+
+	ret = regmap_read(regmap, AXI_PWMGEN_REG_CORE_VERSION, &val);
```
need to check `if (ret)` here.

threexc replied:
Oops, thanks for the catch.

dlech commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,233 @@
+	int ret;
+	u32 val;
+
+	ret = regmap_write(regmap, AXI_PWMGEN_REG_SCRATCHPAD, AXI_PWMGEN_TEST_DATA);
```
Instead of writing and reading arbitrary data to the scratchpad, it could be more useful to read `AXI_PWMGEN_REG_CORE_MAGIC` instead (and compare the value read to `0x504C5347`).

The scratchpad only confirms that we can read/write to that address, but the verifying the magic number means we have the right IP core at this address.

threexc commented:
Pushed a new version that checks the CORE_MAGIC value, and uses the correct value (0x601A3471, not 0x504C5347 as per the wiki). I tried updating the wiki page, but I can't seem to find/access the regmap resource that the page is embedding.

dlech commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,235 @@
+
+	if (val != AXI_PWMGEN_REG_CORE_MAGIC_VAL)
+		return dev_err_probe(dev, -EIO,
+			"failed to read expected value from register: got %x, expected %x\n",
```
```suggestion
			"failed to read expected value from register: got %08x, expected %08x\n",
```

in case there are leading 0s

threexc replied:
fixed

dlech commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -0,0 +1,235 @@
+		return ret;
+
+	if (val != AXI_PWMGEN_REG_CORE_MAGIC_VAL)
+		return dev_err_probe(dev, -EIO,
```
```suggestion
		return dev_err_probe(dev, -ENODEV,
```

not sure it matters much, but the nature of this check is a bit different

threexc replied:
fixed

threexc commented:
Latest push changes axi_pwmgen_apply so that it only does all of the rounding if the device is actually enabled. This was a bug that was causing issues when removing and re-initializing the driver.

threexc commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -160,12 +158,17 @@ static int axi_pwmgen_setup(struct axi_pwmgen_ddata *ddata, struct device *dev)
 	if (ret)
 		return ret;
-
+	
```
Fixed

threexc commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -53,7 +52,7 @@ static const struct regmap_config axi_pwmgen_regmap_config = {
 static struct axi_pwmgen_ddata *to_axi_pwmgen_chip(struct pwm_chip *chip)
 {
-	return container_of(chip, struct axi_pwmgen_ddata, chip);
+	return pwmchip_get_drvdata(chip);
```
Done.

threexc commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -176,24 +179,34 @@ static void axi_pwmgen_clk_rate_exclusive_put(void *data)
 
+	npwm = axi_pwmgen_setup(regmap, dev);
+
+	ddata = devm_kzalloc(dev, sizeof(*ddata), GFP_KERNEL);
```
Thanks for the catch. Fixed.

threexc commented on drivers/pwm/pwm-axi-pwmgen.c:
```diff
@@ -160,12 +158,17 @@ static int axi_pwmgen_setup(struct axi_pwmgen_ddata *ddata, struct device *dev)
+	if (ret)
+		return ret;
+
+	/* Return the numer of PWMs */
```
Done.

dlech commented:
We can probably close this one since discussions are happening on the upstream mailing lists now instead of here.
